### ISSUE TITLE
Multiple hierarchies

### ISSUE THREAD

**Issue by JoFrhwld** [Forced-Alignment-and-Vowel-Extraction/alignedTextGrid#70]
For one textgrid where there are multiple, different hierarchies, both `Top` and `Bottom` will need to be adjusted.

- `Top` will need to have multiple subset classes
- `Bottom` will need to have multiple superset classes

**Comment by chrisbrickhouse** [comment]
I still need to orient myself to the precise operations, but at a high level this sounds like it would move the data structure from being intersecting doubly-linked lists (DLL within tiers and DLL across tiers) to a directed graph (multiple parents, multiple children). Here's a skeleton for that idea.

```python
class Node:
  def __init__(self,next=None,prev=None):
    """ Store arbitrary number of parents (prev) and children (next) """
    self.next=[]
    if type(next) == list:
      self.add_children(next)
    elif next:
      self.add_child(next)
    
    self.prev = []
    if type(prev) == list:
      self.add_parents(prev)
    elif prev:
      self.add_parent(prev)

  # Intuit position in graph and avoid need for special classes
  # Also makes it easier for users to modify node connections
  # on-the-fly, e.g. deleting a tier or fixing an error without
  # opening praat
  def is_bottom(self):
    return( len(self.next) == 0 )

  def is_top(self):
    return( len(self.prev) == 0 )

 def is_singleton(self):
    """ Useful for error detection since we expect no singletons in our graph. """
    return( self.is_top() and self.is_bottom() )

  # Operations for modifying node relationships
  def add_child( self, child ):
    self.next.append(child)
    if self not in child.prev:
      child.add_parent( self )

  def add_children( self, list: children):
    for child in children:
      self.add_child(child)

  def remove_child( self, child ):
    self.next.remove(child)
    if self in child.prev:
      child.prev.remove( self )

  def remove_children( self, list: children ):
    for child in children:
      self.remove_child( child )
  
  # Functions for parents the same but with self.prev
  # ...
  #  Handle node destruction
  def delete( self ):
    for parent in self.prev:
      # For each parent, replace this node with its children
      parent.remove_child( self )
      parent.add_children( self.next )
    for child in self.next:
      # For each child, replace this node with its parents
      child.remove_parent( self )
      child.add_parents( self.prev )
    self.next = []
    self.prev = []
```

Basically, the idea is to allow arbitrary numbers and kinds of parents (and children) so that we can handle nodes (intervals, tiers, etc) being dominated by multiple entities. The downside is that navigation becomes somewhat harder since we don't distinguish types *a priori*. This could be resolved by subclasses implementing a particular on-the-fly search:

```python
class Phone(Node):
  # __init__ etc
  def next_phones( self ):
    return ([ x for x in self.next if type(x) == Phone ])
```

The bigger danger would be graph cycles. I'm not sure how much of a problem that would be, but we'd need to make sure they don't happen (or at least that we detect and handle them). The above implementation is essentially two directed graphs: one that is strictly parent-to-child and another that is strictly child-to-parent. As long as those graphs don't get muddled together and as long as tier relationships are ordered, then we should avoid cycles.

**Comment by JoFrhwld** [comment]
My thought had been to spin off the `Top` and `Bottom` classes to no longer be subclasses of `SequenceInterval`, with special `.set_superset_class()` and `.set_ subset_class()` methods. They wouldn’t need much more than that, since the `Top` and `Bottom` classes are just placeholders, and we never actually create any instances of them.

The one other place this would have a *big* effect would be in the auto-arranging of tiers to be related in `RelatedTiers`, which could be nicer anyway.

https://github.com/Forced-Alignment-and-Vowel-Extraction/alignedTextGrid/blob/6fccb7721085aa01d09c3126859de2ace4b66238/src/aligned_textgrid/sequences/tiers.py#L299-L320

For a more node based approach, something I was eyeing was anytree which has class mixins for node relationships https://anytree.readthedocs.io/en/latest/ Could be useful for outputting graph of interval relationships, but that could also be overkill for this package.

Not quite the same as graph cycles, but one thing I did run into a lot early on was how easy it is to accidentally introduce infinitely recursive relationships when developing & refactoring, which is why I went so far on testing!
